### DISCUSSION TITLE
Similarities with Elmedia Player

### DISCUSSION THREAD

**Discussion by 12joan** [iina/iina#4320] [Q&A] [answered]
Hi,

I noticed that IINA and Elmedia Player have a large number of features in common and have very similar interfaces in a few places, such as the playlist view. My initial assumption was that Elmedia Player was based off IINA, but Electronic Team, Inc. claim that this is not the case, pointing out that Elmedia Player was created earlier than IINA.

Still, there are enough similarities that I would be surprised if this is a coincidence. Were parts of IINA's design intentionally modelled off Elmedia Player?

**Comment by uiryuu** [comment]
I, as one of the main developers of IINA, has never referred that Elmedia player. Actually I've never used it before since it's paid software. Can you provide some screenshots for comparison? Thanks.

**Reply by 12joan** [reply]
Here's a screen recording of all the UI areas I could access without a pro license. You can download the same version I'm using for free from [here](https://www.elmedia-video-player.com/elmedia-player-download.html).

https://user-images.githubusercontent.com/4272090/230103694-7fbd400e-5470-4a40-aa6a-dbcc715222c7.mp4

### Specific similarities

Here are the specific things that are identical or similar:
- Menus
  - The list of top-level menus is very similar
  - File menu
    - Open Online Video (Open URL in New Window)
    - Save Playlist (Save Current Playlist)
  - View menu (Video menu in IINA)
    - Half/Normal/Double Size - Completely identical, including keyboard shortcuts
    - Aspect Ratio
    - Rotate / Mirror / Flip
    - Enter Full Screen / Enter Picture-in-Picture / Float on Top
    - Show Playlist (Show Playlist Panel)
  - Playback
    - Jump to Beginning
    - A-B Loop
    - Next/Previous
    - Increase/Decrease Speed
    - Reset to Normal Speed
    - Take Screenshot
  - Audio
    - Increase/Decrease/Reset Audio Delay
    - Audio Device
  - Subtitles
- The ability to load subtitles from OpenSubtitles.org
- The Media Info (Inspector in IINA) feature looks very similar
- Playlist view is incredibly similar:
  - Overall layout
  - Total duration feature
  - Button placement

**Reply by 12joan** [reply]
I wasn't entirely sure before, but after composing that list, I'm convinced that this can't be a coincidence. It goes beyond one app looking to others like it for features to implement; either IINA or Elmedia must have used the other as its primary source of inspiration.

If one of the developers on IINA took inspiration from Elmedia, this is unlikely to be an issue. Since Elmedia is closed source, there's no risk of anyone infringing the copyright to their source code. Regardless, given the extent of the similarities, using [clean-room design](https://en.wikipedia.org/wiki/Clean_room_design) may have been advisable to prove that no infringement took place.

It's a more serious matter if Elmedia has been copying features from IINA (which they deny). Since IINA is open source, Elmedia would have a difficult time proving that they didn't look at the source code for any of the allegedly copied features. Since IINA is released under GPL, Elmedia may be obliged to publish their source code if parts of it have been copied from IINA in a way that would violate copyright law in the absense of a license. If it comes to this, [the FSF offers help with resolving alleged violations of GPL](https://www.fsf.org/licensing/compliance), which anyone who has contributed code to IINA prior to the alleged copying can instigate, as per IINA's contributing document.

**Reply by svobs** [reply]
I'm not one of the original developers of IINA, though I've been doing a lot of work on it recently. I too have never used Elmedia Player, although I have seen occasional screenshots of it.

After watching the video, I actually came to the conclusion that they just aren't very similar, once you factor out (1) the expected features and their naming conventions, which over time have become standardized, and (2) logical sorting of these features according to their category and functionalities.

From the video above, the player window is completely different than IINA's. The Elmedia window looks more like older versions of QuickTime:
![](https://down10.software/data/download-quicktime-player/screenshot/quicktimexplayer0903071.png)

As for the menu layouts, IINA seems objectively better organized. For example, while IINA places all the playlist-related items under the Playback menu, Elmedia puts some of them under `File`, others under `View`. It does figure out to put `Next`, `Previous`, `Shuffle`, etc. under `Playback`, but as these are playlist-related operations, it would make more sense to group them with the other playlist items like IINA does.

Also, though you compare Elmedia's `View` menu to IINA's `Video` menu, they don't looks the same to me. I see two distinctly different categories which have a lot of overlap.

<img width="244" alt="SCR-20230410-sbav" src="https://user-images.githubusercontent.com/2213815/231051334-dc87bb4f-0662-4adb-8ce5-8d628aef51ac.png"> ![SCR-20230410-sbgz](https://user-images.githubusercontent.com/2213815/231051351-ac9d804c-582c-4c38-97da-b0c2afb262ed.png)

In fact, it looks like the mission or scope of Elmedia's `View` menu is a bit hazy. They put in all the standard functions which change the size, orientation, and behavior of the current window (what I call "behavior": Float on Top, Picture-in-Picture, Full Screen). All of these actions also apply to the video, because the video is the thing filling up most of the window. But IINA is able to include other video-related things such as filters because it is a "Video" menu, whereas Elmedia adds some misc UI things like `Show Controls in a Separate Window` and again, playlist stuff, like it doesn't know where else to put them. (OK, to be fair, I just noticed that for some reason IINA puts `Music Mode` in the `Video` menu. I'm going to declare that to be a bug...)

Same concept for naming. For example, `A-B Loop` is just the standard name for this feature. For a newbie, there's just no quick & easy way to explain what this feature is expected to do. But by using its standard name, users who are familiar with it will immediately recognize it, while users who don't can easily look it up by name. Standardization leads to less friction and a smoother user experience. Same with `Aspect Ratio`, `Rotate`/`Mirror`/`Flip`, etc. And things like `Aspect Ratio` have common values, some more common than others, but plenty in number, and a submenu would be expected per Human Interface Guidelines - either Apple's, or generally, due to the expectations that arise from common norms.

**Reply by 12joan** [reply]
You make some good points, and I'll readilly accept that many of the similarities are coincidental, but surely not all of them can be?

Take the Half/Normal/Double Size menu items, for instance.
<img width="224" alt="image" src="https://user-images.githubusercontent.com/4272090/231079086-60c3bbcc-a6b6-4f46-9b46-7d6f27060108.png">
<img width="495" alt="image" src="https://user-images.githubusercontent.com/4272090/231079133-9d4b4e34-f367-4521-9741-c8f04d5d9b0e.png">

I haven't checked any other video players, but I'd be surprised if there are any other than IINA and Elmedia where:
- It has a feature to double or halve the window size
- That feature exists as menu items "Half Size", "Normal Size", "Double Size", in that order and called exactly that
- The default keyboard shortcuts for those are ⌘0, ⌘1, ⌘2
  - (Unless these shortcuts existed as a standard prior to IINA and Elmedia, they seem too arbitrary to have been arrived at independently)

Same situation with the ⌘], ⌘[, ⌘\ shortcuts to change speed, although those are named differnetly.

**Reply by yuxiqian** [reply]
I think these resize related menu items and keyboard shortcuts originated in QuickTime 7.

<img width="685" alt="QuickTime Pro 7 Keyboard shortcuts" src="https://user-images.githubusercontent.com/34335406/231082759-24d08a6a-10d9-4ddf-9c82-06176d27d01d.png">

**Reply by saagarjha** [reply, marked as answer]
It's very possible that they pulled some inspiration from our UI design, but the code seems substantially different and likely does not share much with IINA.

We can't speak for any of our third-party contributors, but to my knowledge none of the team was aware of this player's existence before it was brought up here. Most of the IINA UI was done by us so it's unlikely that any ideas flowed the other way.
